lib/pos_blockchain.go

```diff
@@ -0,0 +1,1134 @@
+	}
+	return nil
+}
+
+// tryApplyNewTip attempts to apply the new tip to the best chain. It will do the following:
+// 1. Check if we should perform a reorg. If so, it will handle the reorg. If reorging causes an error, return false and error.
+// 2. Check if the incoming block extends the chain tip after reorg. If not, return false and nil
+// 3. If the incoming block extends the chain tip, we can apply it by calling addBlockToBestChain. Return true and nil.
+func (bc *Blockchain) tryApplyNewTip(blockNode *BlockNode, currentView uint64, lineageFromCommittedTip []*BlockNode) (_appliedNewTip bool, _err error) {
+
+	// Check if the incoming block extends the chain tip. If so, we don't need to reorg
+	// and can just add this block to the best chain.
+	chainTip := bc.GetBestChainTip()
+	if chainTip.Hash.IsEqual(blockNode.Header.PrevBlockHash) {
+		bc.addBlockToBestChain(blockNode)
+		return true, nil
+	}
+	// Check if we should perform a reorg here.
+	// If we shouldn't reorg AND the incoming block doesn't extend the chain tip, we know that
+	// the incoming block will not get applied as the new tip.
+	if !bc.shouldReorg(blockNode, currentView) {
+		return false, nil
+	}
+
+	// We need to perform a reorg here. For simplicity, we remove all uncommitted blocks and then re-add them.
+	committedTip, idx := bc.getCommittedTip()
+	if committedTip == nil || idx == -1 {
+		// This is an edge case we'll never hit in practice since all the PoW blocks
+		// are committed.
+		return false, errors.New("tryApplyNewTip: No committed blocks found")
+	}
+	// Remove all uncommitted blocks. These are all blocks that come after the committedTip
+	// in the best chain.
+	// Delete all blocks from bc.bestChainMap that come after the highest committed block.
+	for ii := idx + 1; ii < len(bc.bestChain); ii++ {
+		delete(bc.bestChainMap, *bc.bestChain[ii].Hash)
+	}
+	// Shorten best chain back to committed tip.
+	bc.bestChain = bc.bestChain[:idx+1]
+	// Add the ancestors of the new tip to the best chain.
+	for _, ancestor := range lineageFromCommittedTip {
+		bc.addBlockToBestChain(ancestor)
+	}
+	// Add the new tip to the best chain.
+	bc.addBlockToBestChain(blockNode)
+	return true, nil
+}
+
+// shouldReorg determines if we should reorg to the block provided. We should reorg if
+// this block is proposed in a view greater than or equal to the currentView. Other
+// functions have validated that this block is not extending from a committed block
+// that is not the latest committed block, so there is no need to validate that here.
+func (bc *Blockchain) shouldReorg(blockNode *BlockNode, currentView uint64) bool {
+	chainTip := bc.GetBestChainTip()
+	// If this block extends from the chain tip, there's no need to reorg.
+	if chainTip.Hash.IsEqual(blockNode.Header.PrevBlockHash) {
+		return false
+	}
+	// If the block is proposed in a view less than the current view, there's no need to reorg.
+	return blockNode.Header.ProposedInView >= currentView
```

Comment:
I haven't reviewed the hardcore consensus rules that thoroughly yet. But it looks like this rule makes it so that we'd reorg in the following scenario:
* Suppose `bestChain = A(view=1, committed) -> B(view=2, uncommitted) -> C(view=3, uncommitted)` and now block `A(view=1, committed) -> D(view=4, uncommitted)` comes in, such that D extends from A but with a higher view. In this case I think we would reorg from A->B->C to A->D.

I think the above is all the right thing to do. But just double-check it.

Reply:
That's correct and that's also my understanding of the reorg logic for fast hotstuff. @AeonSw4n @tholonious - this is correct ya?